scheduler/reconcile.go

```diff
@@ -620,7 +620,9 @@ func (a *allocReconciler) cancelUnneededCanaries(original allocSet, desiredChang
 		canaries = all.fromKeys(canaryIDs)
 		untainted, migrate, lost, _, _, _ := canaries.filterByTainted(a.taintedNodes, a.supportsDisconnectedClients, a.now)
 		a.markStop(migrate, "", allocMigrating)
+		desiredChanges.Stop += uint64(len(migrate))
```

Comment:
@tgross While adding my tests, I didn't see the stop count I expected on desired changes so I added these 2 lines, but of course, other tests now fail. Before I refactor those, let me ask. Were we not adding these stops to desiredChanges on purpose? It seemed like an oversight to me because a few lines above other stops by this function do add to the stop count.

Reply:
That's odd. I'm not sure why we'd mark migrating allocs for stop at all. They're left running until we can complete the [`migration`](https://www.nomadproject.io/docs/job-specification/migrate) and _then_ stopped. I would definitely make sure you have tested your understanding of what it's doing currently (and in 1.2.6, tbh... this code has been completely reworked for 1.3 after all) before making further changes here.

Reply:
I did a little digging and this part of this function does look like it has changed since 7/7/17. My understanding of what's happening here is this:

- This function is dealing specifically with canaries that have been placed but not yet finished being verified as healthy.
- Research indicates that once a canary gets marked healthy it won't be in the `PlacedCanaries` but instead will be counted in the `PlacedAllocs` count field.
- I think the intent of this code is similar to how we treat pending allocs when a node disconnects, which is, if it isn't verified running, mark it for stop.
- As far back as 7/5/17, this function is not adding these stop counts to `desiredChanges.Stop`. 
- The earlier version of this function before 7/5/2017 had a slightly better comment - "Handle canaries on migrating/lost nodes here by just stopping them" - so it seems intentional.
- I can see a justification for both incrementing and not incrementing the counter. 
  - On the one hand, a stop is going to get added to the reconciliation result, and work is going to be performed.
  - On the other hand, this is an uncommitted, in-progress long-running transaction. 
  - If you mentally model it as an uncommitted transaction, this is a rollback/abort, which philosophically isn't changing cluster state, because the state was never committed. But in practical terms is work that has to be performed. 
- I guess I'm in favor of keeping existing behavior, reverting these changes, and updating my tests to account for this along with some comments explaining the why.

Thoughts?

Reply:
One last realization.  The lines of the function right above this actually report stops for in-process canaries in the case of a failed deployment or one that is from an older deployment.  I guess we could ask Alex what his intention was but I doubt he'd remember 😄